### ISSUE TITLE
Enable JIT for jailed iOS Users using AltJIT

### ISSUE THREAD

**Issue by lonkle** [jpd002/Play-#1078]
@jpd002 Hi, this is Joelle from the AltStore (iOS) Team. I was told to get ahold of you by @Zer0xFF on here about this - but I'm also on your Discord server so if you want to discuss this with me directly, please feel free to ping me on your server.

Anyway, I wanted to share the very recently completed (like hours ago) library to easily (single function) allow JIT enablement on demand **for jailed iOS users.** Here's the GitHub with the fairly-easy-to-use library needed to integrate into the PPSSPP project: https://github.com/rileytestut/AltKit

And here's my personal opinion on how to use the library to best suit your JIT-needing jailed users: https://trello.com/c/YNjfbtX2/96-create-a-library-for-other-developers-to-use-to-automatically-enable-jit-when-possible

Note: Successful JIT-enabling requires, for a single moment, that the user be on the same WiFi network as any AltServer (which is how jailed users install Play! so they're guaranteed to have AltServer).

**Comment by Spidy123222** [comment]
I did the manual way using libimobiledevice and indeed it works. The method that is proposed should be able to activate it automatically without troubles.

https://user-images.githubusercontent.com/64176728/120909840-2d8de000-c62e-11eb-9bc1-ed5ad0cf3e34.MOV

https://user-images.githubusercontent.com/64176728/120909842-32eb2a80-c62e-11eb-9687-62903bc293ee.MOV

**Comment by jpd002** [comment]
Very nice :) I'm not very familiar with AltStore, I'll take a closer look.

How do you suggest that it should be implement UI side? Should it be a button in the settings? Or something that should be called in the background when the application is started?

**Comment by Spidy123222** [comment]
> Very nice :) I'm not very familiar with AltStore, I'll take a closer look.
> 
> How do you suggest that it should be implement UI side? Should it be a button in the settings? Or something that should be called in the background when the application is started?

it is preferred to do it when the application is started and during it, because it will likely check if it is still debugging. that is all I know of it. extending the debug using location services that are in that Trello card helps it last a lot more hours than without.

**Comment by lonkle** [comment]
@jpd002
• Use the function (shown in the Github readme linked) to ask to enable on JIT on launch
• If it fails at launch you really have **nothing** to lose by requesting JIT at random intervals - given the user's conditions need to be right for the function to respond back to your app with "JIT enabled successful"

And if you don't want to use the function at random intervals **until** JIT gets successful enabled. I'd suggest requesting it once at app launch, and then on each instance the iPhone connects to a WiFi network **and** USB connect.

Does all of that make sense?

**Comment by jpd002** [comment]
Yup, sounds good, thanks! 

Another question, are you planning to have a CMake project for the lib? I'm working on getting something to build with CMake on my end, but would be nice to know if you have something in your pipeline 😄. If I get my thing to work, I could also contribute it to your repo.

**Comment by lonkle** [comment]
@jpd002 Youre the second dev / team that has asked about a Cmake variation of this. The first team was PPSSPP and they said due to their Cmake pipeline they didn't know how to include our library. 

Tbh, I have no idea what Cmake is. So I'm gonna have to do some reading up to see if we can't translate the library. Which shouldn't be hard because the AltKit Library **is not** complex at all. Translating it to a different language / build pipeline (Cmake?) shouldn't be difficult at all. 

If you look at the source for AltKit, you'll see it's actually a very simple "enable" function that isn't too complicated so my impression is that it should be easily made into whatever this cmake pipeline is. But I'm gonna have to learn about this. 

Since both you and PPSSPP use this Cmake tech. I'm assuming it's some cross platform build tech that just makes it easier to write one codebase and it compiles and runs on all platforms it supports?

**Comment by jpd002** [comment]
> I'm assuming it's some cross platform build tech that just makes it easier to write one codebase and it compiles and runs on all platforms it supports?

Yeah, that's basically what CMake is. CMake supports Swift from what I can see, so, compiling the code is not an issue. I'm just having a bit of trouble right now because it doesn't like the module import (AltKit importing CAltKit).

**Comment by jpd002** [comment]
Tried something here: https://github.com/jpd002/AltKit/tree/cmake_poc. The AltKit module can be built properly with that, but then, the next problem is that it's not possible to import the AltKit module inside Objective-C code. 

Seems importing Swift modules relies on C++ module support to work and CMake doesn't support that. I guess we could make it work with some hacks, but I dunno how deep in hacks we're going to be. You can check the `https://github.com/jpd002/Play-/tree/ios_altkit` branch if you'd like to see what I did on the emulator's side.

Would it be possible to have a pure Objective-C library instead of relying on Swift code?

**Comment by lonkle** [comment]
@jpd002 It absolutely can be built that way. iOS 15 **just** got released so we have our hands full at the moment, but I'll reach back out when I can focus on getting you a **purely** Objective-C build of AltKit. Which is great because then we can provide this same solution to the devs of PPSSPP since they also really on CMake. 

Thanks for providing the branch to look at, and telling me what best what suit your pipelines needs. I'll reach back out when we have a solution you can **easily** integrate!

**Comment by lonkle** [comment]
@jpd002 As it turns out, maintaining AltKit would have to be split between Obj-C and Swift if we did made a purely Obj-C package which would take too many resources than we have to maintain them separately.

So I talked about it with the team and this is what we're gonna do. We're building an "Enable JIT and Open" feature directly into AltStore. Here's what we came up with - say a user installs Play! from AltStore, after it finishes installing, they simply have to hit "Open" in our AltStore app and AltStore will **automatically** enable JIT on your app without **any** code changes on your part.

Now, this isn't incredibly ideal on your end given that you have to guide your jailed users to open your app in a specific way (through AltStore instead of directly from the home screen). But it requires no code changes to your build, and allows your jailed users to use JIT. 

Would you maybe like to revisit this as `cmake` gets more mature and potentially in the future will allow for importing `CAltKit` which would mean your users would no longer have to "Open Play! from AltStore for JIT"?

**Comment by jpd002** [comment]
I think it’s an interesting alternative, but I’m not giving up for the CMake solution yet: I’ve done some more research and there might be a nice way to workaround the module import issue. I’ll let you know what happens!

**Comment by lonkle** [comment]
@jpd002 No way! You're a legend. I really thought our hands were gonna be tied on this one till CMake supported that type of import. But if you found a way to do, please tell me so I can write a "AltKit for CMake" instructions guide given you're the only CMake dev I've encountered who has decided to tackle this instead of "wait" (other dev's I'd contacted about `AltKit` previous to you told me that Swift couldn't even be used in CMake so you making **Swift itself** work in CMake was honestly super good news already tbh - I should probs tell those other devs they can do that 🤣).

**Comment by jpd002** [comment]
Alright, I pushed my working version. Seems to compile and run fine, but I don't have all the AltStore setup, so I can't test fully. The build is here if anyone wants to test it: https://github.com/jpd002/Play-/actions/runs/925303034

There seem to be another problem though. The package now weights 150MB instead of a few megabytes. Seems that all Swift libraries are added in the APP file. Do you know if there's a way to reduce that @lonkle?

**Comment by lonkle** [comment]
@Spidy123222 Would you mind testing that out on your system?

@jpd002 As for your final compiled binary issue. I tried looking up resources of CMake and Swift and I wanted to make sure Cmake wasn't set to build your package in "debug mode"?

Other Links I used to try to understand Swift + Cmake Integration:

**Comment by jpd002** [comment]
I might have missed something, I was doing some reading and it seems iOS 12.2 has Swift 5 runtime libraries available: https://swift.org/blog/abi-stability-and-apple/. I'm gonna check if changing the deployment target helps anything.

**Comment by lonkle** [comment]
@jpd002 That's so weird and ridiculous, but I'm sure there's some silly engineer-y reason (there always is), but I don't even care. I'm just happy you were able to integrate our AltKit library into Cmake. That **may just mean** that the PPSSPP team can do so too, I'm gonna link this Issue and tell them to use it as a reference. If there's anything else you can think of to help other devs out also on iOS and using Cmake - just add any info to this ticket since this is going to be my **AltStore Cmake Reference for AltKit**. Thanks so much for all of your work on this, you have no idea how appreciative I am. 🎉

**Comment by jpd002** [comment]
Merged the integration in master. Had some trouble to make this work myself to do my misunderstanding of the process, but I finally made it work on my end 😃 

AltServer integration is enabled with a switch in the settings. It's not enabled by default because I think the "Play! wants to connect to your network to communicate with AltServer" popup is a bit scary for users who start the emulator for the first time. Would probably be nice to have some kind of status indicator saying if AltServer JIT was successfully enabled or not in the Collection view.
